acoustic/flatten_bird_csv/src/flatten_bird_csv.py

```diff
@@ -0,0 +1,179 @@
+Best practice is:
+
+python3.8 -m venv dataflow_env
+. dataflow_env/bin/activate
+pip install apache-beam[gcp]
+"""
+
+import logging
+import os
+from typing import Any, Dict, Iterable, Sequence
+
+from absl import app
+from absl import flags
+from apache_beam.options import pipeline_options
+import apache_beam as beam
+
+FLAGS = flags.FLAGS
+
+flags.DEFINE_string('base_dir', None, 'Parent directory of input/ and output/.')
+flags.DEFINE_string('gcp_project', None,
+                    'Cloud project name; enables DataflowRunner.')
+flags.DEFINE_string('gcp_region', None,
+                    'Cloud region. Required if gcp_project is set.')
+
+AVRO_SCHEMA = {
+    'namespace':
+        'comb',
+    'type':
+        'record',
+    'name':
+        'Prediction',
+    'fields': [
+        {
+            'name': 'key',
+            'type': 'string'
+        },
+        {
+            'name': 'bird',
+            'type': 'string'
+        },
+        {
+            'name': 'begin_seconds',
+            'type': 'float'
+        },
+        {
+            'name': 'logit',
+            'type': 'float'
+        },
+    ]
+}
+
+
+def birds_txt_species_code(line: str) -> str:
+  """Selects the lowercase species code from a line of birds.txt."""
+  parts = line.strip().split('\t')
+  code_parts = parts[1].split('"')
+  assert len(code_parts[0]) == 0
+  assert len(code_parts[2]) == 0
+  return code_parts[1]
+
+
+def flatten(line: str, bird_list: Iterable[str]) -> Iterable[Dict[str, Any]]:
+  """Converts a predictions CSV line to per-species records.
+
+  Args:
+    line: A line from predictions CSV, which has the format (filename, begin,
+      end, logit0, logit1, ...)
+
+  Yields:
+    A dict with the {key, bird, begin_seconds, logit} for each logit on this
+    line.
+  """
+  parts = line.split(',')
+  key = parts[0]
+  begin_seconds = float(parts[1])
+  logits = parts[3:]
+  for code, logit in zip(bird_list, logits):
+    logit = float(logit)
+    yield {
+        'key': key,
+        'bird': code,
+        'begin_seconds': begin_seconds,
+        'logit': logit,
+    }
+
+
+def run(base_dir: str, options: pipeline_options.PipelineOptions) -> None:
+  """Runs the CSV-to-Avro pipeline.
+
+  Args:
+    base_dir: Path to a directory containing input/ and output/ subdirectories.
+      input/birds.txt should have tab-separated values with species codes in the
+      second field. input/ should have any nested directory structure with logit
+      CSV files ending in .csv.
+    options: Runtime pipeline configuration.
+
+  Returns:
+    None
+  """
+  birds_txt = os.path.join(base_dir, 'input', 'birds.txt')
+  csv_pattern = os.path.join(base_dir, 'input', '**', '*.csv')
+  output = os.path.join(base_dir, 'output', 'avro')
+
+  with beam.Pipeline(options=options) as pipeline:
+    bird_list = beam.pvalue.AsSingleton(
+        pipeline | ('ReadBirds' >> beam.io.textio.ReadFromText(birds_txt))
```

Comment:
Do these (presumably parallel-ish) operations guarantee that the output order is maintained between steps?

I would usually just read the bird list before starting the pipeline to ensure that everything is in the correct order. (But I guess that's why I'm not a Haskell programmer.)

Reply:
They do not guarantee order. I knew this but apparently didn't know it. Hmm...

The reason for reading the bird list in the pipeline is not so much a preference to be purely functional as it is that Python open() does not work to read files from GCS. What I should have been doing is using Beam Match transforms to sent Beam ReadableFiles, or their contents, as side inputs.

Reply:
Maybe a good option then would be to use a transform which reads the whole file and then a ParDo to parse the full contents in one go, instead of line-by-line?

Reply:
Hi M & T, 

Looking forward to pulling in this feature, sounds really flexible and powerful.

BTW: See also Mark & my R code from the first 'live' feature for this week:

~/acoustic/data_ingest

https://github.com/calacademy-research/COMB/commit/651a2744bce3e9d351180dba03e04633c1a1b5a5

it gets and puts data from here:

https://drive.google.com/drive/folders/1-BDu2u1T_S_EWdrb25iiroAL_kpWYEmS

including some precomputed logit filters that allow us moRtals to continue to schlep along coding ... ;)

Send your pull request when you are ready! :)

Reply:
Yes, that's kind of what I was getting at with the ReadableFile thing. It worked and is in the later commits.

Reply:
Gotcha; sorry for the noise.